compiler/rustc_mir_build/src/build/coverageinfo/mcdc.rs

```diff
@@ -146,104 +123,663 @@ impl MCDCState {
+                    *next_id = ConditionId::NONE;
+                }
+            });
+        }
+        assert!(
+            start_conditions.remove(&ConditionId::from_usize(1)),
+            "Condition 1 should be always the start"
+        );
+        // Some conditions might not find their predecessors due to gaps between tests.
+        // For instance,
+        // ```rust
+        // (Pat::A(Some(_)), _ ) => { /* ... */ },
+        // (Pat::A(None),    _) => { /* ... */ },
+        // ```
+        // Say `Pat::A` is C1, `Some(_)` is C2, `None` is C3. In the second decision we would
+        // reassign id of `Pat::A` with true next `None` because C2 is not condition of this decision.
+        // `None` becomes "fake start" here as a result of no condition's next linking to it.
+        // Considering condition id of `fake_start` is always less than its first predecessor
+        // due to traverse order in `PatternDecisionCtx::finish`, we can fix these gaps in following way.
+        while let Some(fake_start) = start_conditions.pop_first() {
+            let Some(first_predecessor) = conditions_info.iter_mut().find(|info| {
+                info.condition_id < fake_start && info.true_next_id == ConditionId::NONE
+            }) else {
+                bug!("except condition 1 all conditions must have at least one predecessor");
+            };
+            first_predecessor.true_next_id = fake_start;
+
+            // Note that sibling predecessors probably have larger condition ids.
+            let mut sibling_predecessor = first_predecessor.false_next_id;
+            while sibling_predecessor != ConditionId::NONE {
+                let predecessor_info = conditions_info
+                    .get_mut(sibling_predecessor.as_usize() - 1)
+                    .expect("condition ids are properly assigned with the index");
+                assert_eq!(
+                    predecessor_info.true_next_id,
+                    ConditionId::NONE,
+                    "condition is joint with its false next by '|' in same decision, so they should have same true next"
+                );
+                predecessor_info.true_next_id = fake_start;
+                sibling_predecessor = predecessor_info.false_next_id;
+            }
+        }
+
+        let decision = MCDCDecisionSpan {
+            span: decision_span,
+            decision_depth,
+            conditions_num: conditions.len(),
+            end_markers: vec![],
+        };
+
+        (decision, conditions)
     }
 }
 
+enum DecisionCtx {
+    Boolean(BooleanDecisionCtx),
+    Pattern(PatternDecisionCtx),
+}
+
 pub struct MCDCInfoBuilder {
-    branch_spans: Vec<MCDCBranchSpan>,
-    decision_spans: Vec<MCDCDecisionSpan>,
-    state: MCDCState,
+    normal_branch_spans: Vec<MCDCBranchSpan>,
+    mcdc_spans: Vec<(MCDCDecisionSpan, Vec<MCDCBranchSpan>)>,
+    decision_ctx_stack: Vec<DecisionCtx>,
+    decision_ends: FxIndexMap<Span, Rc<RefCell<Vec<BlockMarkerId>>>>,
```

Comment:
`decision_ends` is introduced because:
1. Candidates with `|` may not have `pre_binding_block` when matching finishes.
2. For every arm in `match` statements, `arm_block` is generated in `lower_match_arms` and immediately compiler digs into code in the arm.
Hence when a `PatternDecisionCtx` is finished, we do not know all blocks which are "ends" of these decisions. Instead we fulfill them in `add_ends_to_decision` later. Because decisions produced by same `ctx` share same ends, `Rc<RefCell<_>>` is used here.